nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/OAuth2Object.java

```diff
@@ -188,54 +192,195 @@ public Response getToken(@PathParam("providerId") String providerId, @Context Ht
+        return updateProviderToken(providerId, nxuser, request, token);
+    }
+
+    /**
+     * Deletes an OAuth2 provider token.
+     *
+     * @since 10.2
+     */
+    @DELETE
+    @Path("token/provider/{providerId}/user/{nxuser}")
+    public Response deleteProviderToken(@PathParam("providerId") String providerId, @PathParam("nxuser") String nxuser,
+                                        @Context HttpServletRequest request) {
+        checkPermission(nxuser);
         NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
-        return Response.ok(updateToken(provider, nxuser, token)).build();
+        deleteToken(getTokenDoc(provider, nxuser));
+        return Response.noContent().build();
     }
 
     /**
      * Deletes an OAuth2 Token.
      *
      * @since 9.2
+     *
+     * @deprecated since 10.2 Use {@link #deleteProviderToken(String, String, HttpServletRequest)} instead.
      */
+    @Deprecated
     @DELETE
     @Path("token/{providerId}/{nxuser}")
     public Response deleteToken(@PathParam("providerId") String providerId, @PathParam("nxuser") String nxuser,
-            @Context HttpServletRequest request) {
-        checkPermission();
-        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
-        deleteToken(getTokenDoc(provider, nxuser));
+                                @Context HttpServletRequest request) {
+        return deleteProviderToken(providerId, nxuser, request);
+    }
+
+    /**
+     * Retrieves all oauth2 provider tokens for the current user.
+     *
+     * @since 10.2
+     */
+    @GET
+    @Path("token/provider")
+    public List<NuxeoOAuth2Token> getProviderUserTokens(@Context HttpServletRequest request) {
+        String nxuser = request.getUserPrincipal().getName();
+        return getTokens(nxuser).stream() // filter: make sure no client tokens are retrieved
+                .filter(token -> token.getClientId() == null).collect(Collectors.toList());
+    }
+
+    /**
+     * Retrieves all oauth2 client tokens for the current user.
+     *
+     * @since 10.2
+     */
+    @GET
+    @Path("token/client")
+    public List<NuxeoOAuth2Token> getClientUserTokens(@Context HttpServletRequest request) {
+        String nxuser = request.getUserPrincipal().getName();
+        return getTokens(nxuser).stream() // filter: make sure no provider tokens are retrieved
+                .filter(token -> token.getClientId() != null).collect(Collectors.toList());
+    }
+
+    /**
+     * Retrieves a oauth2 client token.
+     *
+     * @since 10.2
+     */
+    @GET
+    @Path("token/client/{clientId}/user/{nxuser}")
+    public Response getClientToken(@PathParam("clientId") String clientId, @PathParam("nxuser") String nxuser,
+                              @Context HttpServletRequest request) {
+        checkPermission(nxuser);
+        OAuth2Client client = Framework.getService(OAuth2ClientService.class).getClient(clientId);
```

Comment:
WDYT about returning a 404 for a non-existing client as it's done for providers? 
If that happens, we will end up with a 5xx but that's not a server issue at all.

Reply:
Thanks. I agree, and I will do the same for getProviderToken as well.